builtin/common/serialize.lua

```diff
@@ -1,205 +1,190 @@
-			local vals = {}
-			local idx_dumped = {}
-			local np = nest_points[x]
-			for i, v in ipairs(x) do
-				if not np or not np[i] then
-					vals[#vals + 1] = dump_or_ref_val(v)
-				end
-				idx_dumped[i] = true
-			end
-			for k, v in pairs(x) do
-				if (not np or not np[k]) and
-						not idx_dumped[k] then
-					vals[#vals + 1] = "["..dump_or_ref_val(k).."] = "
-						..dump_or_ref_val(v)
-				end
-			end
-			return "{"..table.concat(vals, ", ").."}"
-		else
-			error("Can't serialize data of type "..tp)
+local function serialize(value, write)
+    local reference = {"A"}
+    local function increment_reference(place)
+        if not reference[place] then
+            reference[place] = "B"
+        elseif reference[place] == "Z" then
+            reference[place] = "A"
+            return increment_reference(place + 1)
+        else
+            reference[place] = succ[reference[place]]
+        end
+    end
+    local references = {}
+    local to_fill = {}
+    for object, count in pairs(count_objects(value)) do
+        local type_ = type(object)
+        if count >= 2 and (type_ ~= "string" or #reference + 2 < #object) then
+            local ref = table_concat(reference)
+            write(ref)
+            write"="
+            write(type_ == "table" and "{}" or quote(object))
+            write";"
+            references[object] = ref
+            if type_ == "table" then
+                to_fill[object] = ref
+            end
+            increment_reference(1)
+        end
+    end
+    local function is_short_key(key)
+        return not references[key] and type(key) == "string" and string_match(key, "^[%a_][%a%d_]*$")
+    end
+    local function dump(value)
+        -- Primitive types
+        if value == nil then
+            return write"nil"
+        end
+        if value == true then
+            return write"true"
+        end
+        if value == false then
+            return write"false"
+        end
+        local type_ = type(value)
+        if type_ == "number" then
+            return write(string_format("%.17g", value))
+        end
+		if type_ == "function" then
+			return write(string_format("loadstring(%q)", string_dump(value)))
 		end
-	end
-
-	local function dump_nest_points()
-		for parent, vals in pairs(nest_points) do
-			for k, v in pairs(vals) do
-				local_defs[#local_defs + 1] = dump_or_ref_val(parent)
-					.."["..dump_or_ref_val(k).."] = "
-					..dump_or_ref_val(v)
-			end
-		end
-	end
-
-	mark_multiple_occurences(x)
-	local top_level = dump_or_ref_val(x)
-	dump_nest_points()
+        -- Reference types: table and string
+        local ref = references[value]
+        if ref then
+            return write(ref)
+        elseif type_ == "string" then
+            return write(quote(value))
+        elseif type_ == "table" then
+            local first = true
+            write"{"
+            local len = #value
+            for i = 1, len do
+                if not first then write";" end
```

Comment:
Why not terminate it after each value, rather than checking for first occurrences?

Reply:
> Why not terminate it after each value, rather than checking for first occurrences?

Because I don't like trailing semicolons/commas. But it is an option. Either way, performance impact is pretty negligible.